Linux-Device-Driver/2_Character_Driver/README.md

```diff
@@ -0,0 +1,297 @@
+- Write application running on user-level for test the driver functionality.
+
+## 2. Connection establishment between Device file access and the Driver
+
+### Step 1: Create device number
+- The device number is a combination of major and minor numbers.
+- In Linux kernel, `dev_t` *(typedef of u32)* type is used to represent the device number.
+- Out of **32 bits**, **12 bits** to store **major number** and remaining **20 bits** to store **minor number**
+- You can use the below macros to extract major and minor parts of `dev_t` type variable:
+```text
+#include <linux/kdev_t.h>
+
+dev_t device_number;
+int minor = MINOR(device_number);
+int major = MAJOR(device_number);
+```
+- If you have major and minor numbers, use the below macro to turn them into `dev_t` type device number:
+```text
+MKDEV(int major, int minor);
+```
+**- Dynamically register a range of char device numbers**
+```text
+int alloc_chrdev_region(dev_t *dev, unsigned baseminor, unsigned count, const char *name)
+```
++ **dev_t \*dev**: Output parameter for first assigned number 
++ **unsigned baseminor**: first of the requested range of minor numbers
++ **unsigned count**: number of minor numbers required
++ **const char \*name**: name of the associated device or driver
+
+**Example:**
+```text
+#include <linux/fs.h>
+
+dev_t device_number;
+alloc_chrdev_region(&device_number, 0, 1, "pcd_devices");
+```
+
+### Step 2: Char Device Registration with VFS
+**2.1 - Initialize a cdev_init structure**
+```text
+void cdev_init (struct cdev *cdev, const struct file_operations *fops)
+```
++ **struct cdev \*cdev**: the structure to initialize
++ **const struct file_operations \*fops**: the file operations for this device (open, read, write, lseek, mmap, flush,...)
++ `THIS_MODULE` is a macro which resolves in to "pointer to a struct module variable which corresponds to our current module"
+
+**2.2 - add a char device(pcd_cdev) to the system**
+```text
+int cdev_add (struct cdev *p, dev_t dev, unsigned count)
+```
++ **struct cdev \*p**: the `cdev` structure for the device
++ **dev_t dev**: the first device number for which this device is responsible
++ **unsigned count**: the number of consecutive minor numbers corresponding to this device
+
+**Example:**
+```text
+#include <linux/cdev.h>
+
+struct cdev pcd_cdev;
+struct file_operations pcd_fops = {
+    .open       = pcd_open,
+    .write      = pcd_write,
+    .read       = pcd_read,
+    .release    = pcd_release,
+    .llseek     = pcd_lseek,
+    .owner      = THIS_MODULE
+};
+
+cdev_init(&pcd_cdev, &pcd_fops);
+cdev_add(&pcd_cdev, device_number, 1);
+```
+
+## `Kernel Modules exposed device information to sysfs ......`
+## `User Space i.e udev takes that information from sysfs`
+
+## `uevents are generated when Device Driver takes help of kernel API to dynamic creation of the Device File`
+
+### Step 3: Create Device Files
+
```

Comment:
Can we remove this redundant image dir?

Reply:
We can do it but not important we can delete it after merging.